pkg/k8s/certificate.go

```diff
@@ -18,81 +18,198 @@ import (
 	"context"
 	"fmt"
 	"io/ioutil"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/kubernetes"
 	"os"
 	"path"
+	"strconv"
+	"strings"
 
 	log "github.com/sirupsen/logrus"
 	"k8s.io/apimachinery/pkg/api/errors"
 
 	"github.com/tigera/key-cert-provisioner/pkg/cfg"
 	"github.com/tigera/key-cert-provisioner/pkg/tls"
 
-	"k8s.io/api/certificates/v1beta1"
+	certV1 "k8s.io/api/certificates/v1"
+	certV1beta1 "k8s.io/api/certificates/v1beta1"
 	v1 "k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
-// WatchCSR Watches the CSR resource for updates and writes results to the certificate location (which should be mounted as an emptyDir)
+type VersionInfo struct {
+	Minor int
+}
+
 func WatchCSR(ctx context.Context, restClient *RestClient, cfg *cfg.Config, x509CSR *tls.X509CSR) error {
-	watcher, err := restClient.Clientset.CertificatesV1beta1().CertificateSigningRequests().Watch(ctx, metav1.ListOptions{})
+	version, err := getKubernetesVersion(restClient.Clientset)
+	if err != nil {
+		return err
+	}
+
+	watcher, err := createCSRWatcher(ctx, restClient, version, err)
 	if err != nil {
 		return fmt.Errorf("unable to watch certificate requests: %w", err)
 	}
 	log.Infof("watching CSR until it has been signed and approved: %v", cfg.CSRName)
-	for event := range watcher.ResultChan() {
-		chcsr, ok := event.Object.(*v1beta1.CertificateSigningRequest)
+	return WatchCSRBasedOnKubernetesVersion(watcher, cfg, x509CSR, version)
+}
+
+func createCSRWatcher(ctx context.Context, restClient *RestClient, version *VersionInfo, err error) (*watch.Interface, error) {
+	var watcher watch.Interface
+	if version != nil && version.Minor > 18 {
+		watcher, err = restClient.Clientset.CertificatesV1().CertificateSigningRequests().Watch(ctx, metav1.ListOptions{})
+	} else {
+		watcher, err = restClient.Clientset.CertificatesV1beta1().CertificateSigningRequests().Watch(ctx, metav1.ListOptions{})
+	}
+	return &watcher, err
+}
+
+func WatchCSRBasedOnKubernetesVersion(watcher *watch.Interface, cfg *cfg.Config, x509CSR *tls.X509CSR, version *VersionInfo) error {
```

Comment:
It seems like there is no need to export this func as it  is only used within this file.